README.md
Redis Module for Puppet
=======================
[![Build Status](https://secure.travis-ci.org/fsalum/puppet-redis.png)](http://travis-ci.org/fsalum/puppet-redis)

This module installs and manages a Redis server. All redis.conf options are
accepted in the parameterized class.

Important
---------

If you are upgrading this module from 0.x to 1.0+, please test it carefully 
outside production as it is not fully backwards compatible.

Some class parameters were added, removed or had their default values changed.

The redis.conf template has been completely rewritten to support Redis 2.2+ to 2.8+.

Operating System
----------------

Tested on CentOS 6.5, Ubuntu Saucy/Trusty/Precise, Debian 7.4  
redis.conf options compatible with Redis 2.2, 2.4, 2.6, 2.8  

Quick Start
-----------

Use the default parameters:

    class { 'redis': }

To change the port and listening network interface:

    class { 'redis':
      conf_port => '6379',
      conf_bind => '0.0.0.0',
    }

Parameters
----------

Check the [init.pp](https://github.com/fsalum/puppet-redis/blob/master/manifests/init.pp) file for a complete list of parameters accepted.

* custom sysctl

To enable and set important Linux kernel sysctl parameters as described in the [Redis Admin Guide](http://redis.io/topics/admin) - use the following configuration option:

    class { 'redis':
      system_sysctl => true
    }

By default, this sysctl parameter will not be enabled. Furthermore, you will need the sysctl module defined in the [Modulefile](https://github.com/fsalum/puppet-redis/blob/master/Modulefile) file.

* service restart

If you need to execute a controlled restart of redis after changes due master/slave relationships to avoid that both are restarted at the same time use the parameter below.

    class { 'redis':
      service_restart => false
    }

By default service restart is true.


Sentinel
--------

This module supports Redis Sentinel that comes with Redis 2.8+ with all the configuration parameters. It doesn't install Sentinel, the redis-sentinel binary must be already configured, but it manages upstart scripts (can be deactivated with parameter manage_upstart_scripts = false).

Example:

    class { redis::sentinel:
      conf_port      => '26380',
      conf_dir       => '/mydir',
      sentinel_confs => {
        'mymaster' => {
          'monitor'                 => '127.0.0.1 6379 2',
          'down-after-milliseconds' => '60000',
          'notification-script'     => '/etc/redis/scripts/thescript.py',
          'parallel-syncs'          => '3',
        }
        'resque' => {
          'monitor'                 => 'resque 6379 4',
          'down-after-milliseconds' => '10000',
          'failover-timeout'        => 180000,
          'notification-script'     => '/etc/redis/scripts/thescript.py',
          'parallel-syncs'          => '5',
        }
      }
    }


Copyright and License
---------------------

Copyright (C) 2012 Felipe Salum

Felipe Salum can be contacted at: fsalum@gmail.com

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG
## fsalum-redis changelog

Release notes for the fsalum-redis module.

------------------------------------------

#### 1.0.2 - 2014-10-25

* Fixed Travis integration and tests
* Added metadata.json
* Making it compatible to be 'puppet approved' :)

#### 1.0.1 - 2014-09-16

* Allowing redis package name as a param. #35 (@nprimmer)
* add $redis_version_override #37 (@tmclaugh)

#### 1.0.0 - 2014-06-08

Many CHANGES to this version, complete rewrite of redis.conf template 
to support Redis 2.2 to 2.8.

Make sure to test the module and parameters before upgrading in production.

Thanks to @zxjinn and @zeroecco for their hard work on this release.

* some parameters were removed, added and/or default values changed
* update redis.conf to the latest version available #32 (@zxjinn)
* Logic for redis config file to support 2.2.x through 2.8.x #31 (@zeroecco)
* Unixsocketoptions #33 (@nbeernink)
* Changed operating system check to validate osfamily #29 (@george-b)

#### 0.0.12 - 2014-03-21

* Unset cleanup #27 (@trlinkin)
* toggle ability to notify service to restart when config file changes #28 (@tmclaugh)

#### 0.0.11 - 2014-02-19

* system_sysctl parameter for redis class (@tehmaspc)

#### 0.0.10 - 2014-02-19

* Allow conf_bind to be unset (@stevelacey)
* Changing default of glueoutputbuf to UNSET (@tehmaspc)

#### 0.0.9 - 2014-01-09

* Add quotes around all instances of UNSET (@charlesdunbar)

#### 0.0.8 - 2013-12-20

* Setting mode permission for conf_dir (@KlavsKlavsen)

#### 0.0.7 - 2013-12-11

* Add glueoutputbuf config option (@kryptx)
* Add support for Amazon Linux AMI (@mattboston)

#### 0.0.6 - 2013-08-01

* Install package before poking config (@doismellburning)

#### 0.0.5 - 2013-06-06

* Fix Puppet 3.2.1 deprecation warnings (@ripienaar)
* Fix duplicate entry for logrotate on Debian/RHEL using different paths (@arthurfurlan)
* Add $conf_nosave parameter (@fsalum)
* Minor changes to params variables (@fsalum)
* Update CHANGELOG format

------------------------------------------

#### 0.0.4 - 2013-04-18

* Creates conf_dir directory if it doesn't exist

------------------------------------------

#### 0.0.3 - 2013-02-25

* Fixing redis.conf less options on Debian

------------------------------------------

#### 0.0.2 - 2013-02-25

* Fixing redis.conf location for Debian

------------------------------------------

#### 0.0.1 - 2013-02-25

* Initial Forge release


spec/classes/redis_spec.rb
# for rspec-puppet documentation - see http://rspec-puppet.com/tutorial/
require_relative '../spec_helper'

describe 'redis' do
  let(:facts) { {:operatingsystem  => 'ubuntu'} }

  it { should contain_class('redis::params') }

  it do
    should contain_package('redis').with(
      'ensure'  =>  'present',
      'name'    =>  'redis-server',
    )
  end

  it do
    should contain_service('redis').with(
      'ensure'      =>  'running',
      'name'        =>  'redis-server',
      'enable'      =>  'true',
      'hasrestart'  =>  'true',
    )
  end

  it do
    should contain_file('/etc/redis/redis.conf').with(
      'path'  =>  '/etc/redis/redis.conf',
      'owner' =>  'root',
      'group' =>  'root',
      'mode'  =>  '0644',
    )
  end

  it do
    should contain_file('/etc/logrotate.d/redis-server').with(
      'path'  =>  '/etc/logrotate.d/redis-server',
      'owner' =>  'root',
      'group' =>  'root',
      'mode'  =>  '0644',
    )
  end

  it do
    should contain_exec('/var/lib/redis/').with(
      'path'    =>  '/bin:/usr/bin:/sbin:/usr/sbin',
      'command' =>  'mkdir -p /var/lib/redis/',
      'user'    =>  'root',
      'group'   =>  'root',
      'creates' =>  '/var/lib/redis/',
    )
  end

  it do
    should contain_file('/var/lib/redis/').with(
      'ensure' =>  'directory',
      'owner'  =>  'redis',
      'group'  =>  'redis',
      'mode'   =>  '0755',
    )
  end

end


serverspec/spec/redis_spec.rb
# For serverspec documentation - see http://serverspec.org/tutorial.html
require_relative '../spec_helper'

describe package('redis-server') do
  it { should be_installed }
end

describe service('redis-server') do
  it { should be_enabled }
  it { should be_running }
end

describe file('/etc/redis/redis.conf') do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should be_mode '644' }
end

describe file('/etc/logrotate.d/redis-server') do
  it { should be_file }
  it { should be_owned_by 'root' }
  it { should be_grouped_into 'root' }
  it { should be_mode '644' }
end

describe file('/var/lib/redis/') do
  it { should be_directory }
  it { should be_owned_by 'redis' }
  it { should be_grouped_into 'redis' }
  it { should be_mode '755' }
end

describe command('redis-cli --version') do
  it { should return_exit_status 0 }
  it { should return_stdout /^redis-cli.*/ }
end

describe linux_kernel_parameter('vm.overcommit_memory') do 
  its(:value) { should eq 1 }
end


lib/facter/redis_version.rb
require 'facter'

Facter.add("redis_version", :timeout => 120) do
    confine :osfamily => "Debian"

    setcode do

        dpkg = `which apt-cache 2> /dev/null`.chomp
        if dpkg == ''
            dpkg = '/usr/bin/apt-cache'
        end

        redis_version = Facter::Util::Resolution.exec('/usr/bin/redis-server --version')
        if redis_version.nil?
            redis_version = Facter::Util::Resolution.exec(dpkg+" show redis-server 2> /dev/null | /bin/grep -i 'version:' | /usr/bin/awk '{print $2}'").strip
        end

        case redis_version
            when /2\.8\.[0-9]/
                #set version to 2.8
                redis_version = '2.8.x'
            when /2\.6\.[0-9]/
                #set version to 2.6
                redis_version = '2.6.x'
            when /2\.4\.[0-9]/
                #set version to 2.4
                redis_version = '2.4.x'
            when /2\.2\.[0-9]/
                #set version to 2.2
                redis_version = '2.2.x'
            else
                redis_version = 'nil'
        end
        redis_version
    end
end

Facter.add("redis_version", :timeout => 120) do
    confine :osfamily => "RedHat"

    setcode do

        yum = `which yum 2> /dev/null`.chomp
        if yum == ''
            yum = '/usr/bin/yum'
        end

        redis_version = Facter::Util::Resolution.exec('/usr/sbin/redis-server --version')
        if redis_version.nil?
            redis_version = Facter::Util::Resolution.exec(yum+" info redis 2> /dev/null | /bin/grep '^Version' | /bin/awk -F ':' '{print $2}'").strip
        end

        case redis_version
            when /2\.8\.[0-9]/
                #set version to 2.8
                redis_version = '2.8.x'
            when /2\.6\.[0-9]/
                #set version to 2.6
                redis_version = '2.6.x'
            when /2\.4\.[0-9]/
                #set version to 2.4
                redis_version = '2.4.x'
            when /2\.2\.[0-9]/
                #set version to 2.2
                redis_version = '2.2.x'
            else
                redis_version = 'nil'
        end
        redis_version
  end
end
